### ISSUE TITLE
scrape: Bail out of invalid scrape format rather than silently falling back to the classic text format

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#13214]
### Proposal

Back in the old days, it was a reasonable assumption that a target that fails to properly negotiate the scrape format is probably an ad-hoc implementation that most likely tries to produce the (classic) text format. Therefore, it was a good idea to fall back to parsing the classic text format whenever anything went wrong.

However, this assumption doesn't hold anymore. It might even be dangerous. That's mostly because now there is OpenMetrics, which looks very similar to the classic Prometheus text format but is not the same. With the current fallback behavior, Prometheus might end up in a situation where it parses OpenMetrics as the classic text format, which - in the worst case - works without error but yielding wrong results (most relevant when it comes to timestamps).

A recent issue caused by this is #13172. In short, the following happened: With native histograms enabled, Prometheus tries to negotiate protobuf. However, [KSM](https://github.com/kubernetes/kube-state-metrics) wasn't prepared for that and responded with some invalid content type (details to be investigated), which caused Prometheus to parse the KSM scrape as classic text format, while KSM was serving OpenMetrics. Luckily, the metrics exposed by KSM use features of OM that won't parse as the classic text format, so Prometheus error'd out rather than silently ingesting wrong data. However, the error message is still quite confusing (see #13272) and it would have been much clearer if Prometheus had just reported that KSM wasn't negotiating a valid content type.

So the proposal here is to simply not do a fallback but report an error in case of invalid content negotiation.
(Relevant code: https://github.com/prometheus/prometheus/blob/b8bcaef14da44eb384e8a07655da35614dc97467/scrape/scrape.go#L1404-L1410 https://github.com/prometheus/prometheus/blob/b8bcaef14da44eb384e8a07655da35614dc97467/model/textparse/interface.go#L81-L91 )

This is a breaking change as currently there might be a lot of targets that aren't doing proper content negotiation but "just work". This change would break those use cases, but with the danger of ingesting wrong data (as explained above), we should make use of the 3.0.0 release to deliberately expose those broken scrapers.

An open question is if we want to retain the simplicity of doing trivial text-format expositions without _any_ kind of content negotiation, i.e. the [case where the content type ends up empty](https://github.com/prometheus/prometheus/blob/b8bcaef14da44eb384e8a07655da35614dc97467/model/textparse/interface.go#L84-L86). OpenMetrics has created a dilemma here. The subtle differences between OM and the classic text format create ambiguities (see above). We could retain the feature of using one specific format in case the target doesn't do any content negotiation. But which format would that be? We have only bad choices here:
1. If we stuck with the classic text format, the existing use cases won't break, but we would extend the life of the deprecated classic text format forever.
2. If we fell back to OpenMetrics, we would run into the danger of ingesting old targets in the wrong way (most relevant concern: timestamps).
3. We could only accept a "safe common subset" of OM and the classic format in case of missing content negotiation. This would only break a few use cases, but it would be a weird and complicated contraption to cater for a supposedly "trivial" use case.

_(Edited for clarity and correctness. Previously I assumed, Prometheus would also fall back on parsing problem when one content type was negotiated but another one was served, but it doesn't.)_

**Comment by bwplotka** [comment]
Thanks for raising this.

> However, this assumption doesn't hold anymore. It might even be dangerous. That's mostly because now there is OpenMetrics, which looks very similar to the classic Prometheus text format but is not the same. With the current fallback behavior, Prometheus might end up in a situation where it parses OpenMetrics as the classic text format, which - in the worst case - works without error but yielding wrong results (most relevant when it comes to timestamps).

Agree, but wonder.. can we finally do the improvements to OM text we always wanted and perhaps "revert" or change some ideas which were not ideal at the end (e.g _created series and perhaps timestamp change?) and release OM 2.0?

Alternatively we could add some required marker for OM? Or perhaps there is way for OM 2.0 to be... backward compatible with Prometheus text (that would be PARADISE 🙈 ).

Maybe I am looking too wide here, but it feels whatever we do here is wrong, without changing OM a bit. But it feels we version OM for a reason, so why being blocked here?

**Comment by beorn7** [comment]
Thinking about this a bit more as we just discussed it during the dev summit:

We (IIRC) pretty much agreed with the conclusions and actions proposed here. However, I am now even more concerned about breaking the "simple" use case of not negotiating anything and then falling back to the classic text format. I have enumerated three different options to address it above, but I'm not entirely happy with any of those. On the other hand, I think I would be even less happy with not addressing the fallback case at all. So I would like to add a 4th idea:

Let's add a scrape config option to specify a fall-back format (next to the setting to configure content negotiation preferences). That's another knob added, but I think it makes sense because we can then error out by default, exposing the ill-defined cases as described above, and then users can either fix their content negotiation, or – if that's not an option – they can at least explicitly state what the fallback should be.
